RavenDB-16614 Atomic read write in cluster wide transaction without c…

ayende opened the pull request:
…onflicts.

* Cluster wide transactions now will default to creating atomic guards (compare exchange values) that match the modified documents in the transaction. This will ensure that all the documents are modified seamlessly across multiple concurrent cluster wide transactions.
* This behavior can be disabled using DisableAtomicDocumentWritesInClusterWideTransaction flag at the document store, session and database/server levels, but is otherwise on by default.
* atomic compare exchange values has the format: rvn-atomic/{document-id} and contains a reference their owner document.
* Created ClusterTransactionId property in the DatabaseRecord, which is used for database group key for cluster wide transaction index. This can be used to compare to atomic guards, compare exchange values that need to match the same index.
* The ClusterTransactionId is used to enhance change vectors of cluster wide transactions. You can identify them using the TXRN tag.
* @refresh and @expires now handle failed concurrency and conflict errors better
* On import, we'll reset the relevant change vector to match the appropriate compare exchange values as well as remove any orphan guards
* Standard transactions also check whatever the document was modified in a cluster wide transaction and prevent saving documents with the wrong cluster transaction index, which should alleviate issues with mismatched versions
* Such checks do not apply when accepting data via replication, to prevent a failure of transaction matches from breaking replication between nodes.
* The client will now send the OldChangeVector always to the server. This is used to check whatever the document is modified from an up to date version compared to the cluster transaction index.

Misc:
* Fixed metadata save of complex objects
* Fixed concurrent collection usage on Candidate
* Fixed test failure when using WellKnownCertificates as environment variables

ayende committed:
RavenDB-16614 Atomic read write in cluster wide transaction without conflicts.
* Cluster wide transactions now will default to creating atomic guards (compare exchange values) that match the modified documents in the transaction. This will ensure that all the documents are modified seamlessly across multiple concurrent cluster wide transactions.
* This behavior can be disabled using DisableAtomicDocumentWritesInClusterWideTransaction flag at the document store, session and database/server levels, but is otherwise on by default.
* atomic compare exchange values has the format: rvn-atomic/{document-id} and contains a reference their owner document.
* Created ClusterTransactionId property in the DatabaseRecord, which is used for database group key for cluster wide transaction index. This can be used to compare to atomic guards, compare exchange values that need to match the same index.
* The ClusterTransactionId is used to enhance change vectors of cluster wide transactions. You can identify them using the TXRN tag.
* @refresh and @expires now handle failed concurrency and conflict errors better
* On import, we'll reset the relevant change vector to match the appropriate compare exchange values as well as remove any orphan guards
* Standard transactions also check whatever the document was modified in a cluster wide transaction and prevent saving documents with the wrong cluster transaction index, which should alleviate issues with mismatched versions
* Such checks do not apply when accepting data via replication, to prevent a failure of transaction matches from breaking replication between nodes.
* The client will now send the OldChangeVector always to the server. This is used to check whatever the document is modified from an up to date version compared to the cluster transaction index.

Misc:
* Fixed metadata save of complex objects
* Fixed concurrent collection usage on Candidate
* Fixed test failure when using WellKnownCertificates as environment variables

arekpalinski commented on src/Raven.Client/Documents/Conventions/DocumentConventions.cs:
```diff
@@ -1124,6 +1125,20 @@ public void RegisterQueryMethodConverter(QueryMethodConverter converter)
+        /// EXPERT: When the TransactionMode is 'ClusterWide', will disable atomic document writes and validate only compare exchange
+        /// values that are manually added to the user.
+        /// </summary>
+        public bool? DisableAtomicDocumentWritesInClusterWideTransaction
```
"added _to_ the user"?

arekpalinski replied:
I found the description of `Cluster.DisableAtomicDocumentWrites` server side configuration option way easier to understand:

"EXPERT: Disable automatic atomic writes with cluster write transactions. If set to 'true', will only consider explicitly added compare exchange values to validate cluster wide transactions."

arekpalinski commented on src/Raven.Client/Documents/Commands/Batches/DeleteCommandData.cs:
```diff
@@ -9,24 +9,38 @@ namespace Raven.Client.Documents.Commands.Batches
     public class DeleteCommandData : ICommandData
     {
         public DeleteCommandData(string id, string changeVector)
+            :this(id, changeVector, changeVector)
```
Shouldn't we mark this ctor as obsolete?

ayende replied:
I'm worried about users that are using that explicitly.
We encourage that when using `Defer`, so that is part of the interface that we care about.
`OldChangeVector` is mostly not something that I expect users to really notice or care about.

arekpalinski replied:
OK - makes sense

arekpalinski commented on src/Raven.Client/Documents/Commands/Batches/DeleteCommandData.cs:
```diff
@@ -9,24 +9,38 @@ namespace Raven.Client.Documents.Commands.Batches
         public string Name { get; } = null;
         public string ChangeVector { get; }
         public CommandType Type { get; } = CommandType.DELETE;
+        public BlittableJsonReaderObject Document { get; set; }
```
Why is it called "Document" while we're assigning document metadata to it?

ayende replied:
No longer needed, removed.

arekpalinski commented on src/Raven.Client/Documents/Commands/Batches/DeleteCommandData.cs:
```diff
@@ -9,24 +9,38 @@ namespace Raven.Client.Documents.Commands.Batches
                 [nameof(ChangeVector)] = ChangeVector,
-                [nameof(Type)] = Type.ToString()
+                [nameof(Type)] = Type.ToString(),
+                [nameof(Document)] = Document
```
Can we add it to the json only if it's not null? Most of the time it's going to be null so no need to send additional property over the wire

arekpalinski commented on src/Raven.Client/Documents/Session/SessionOptions.cs:
```diff
@@ -23,5 +23,11 @@ public class SessionOptions
+
+        /// <summary>
+        /// EXPERT: When the TransactionMode is 'ClusterWide', will disable atomic document writes and validate only compare exchange
+        /// values that are manually added to the user.
```
Same comment as above

arekpalinski commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -54,11 +56,44 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
+
+            public void ValidateAtomicGuard(string id, NonPersistentDocumentFlags nonPersistentDocumentFlags, string changeVector)
+            {
+                if (nonPersistentDocumentFlags != NonPersistentDocumentFlags.None) // replication or engine running an operation, we can skip checking it 
```
`NonPersistentDocumentFlags` has _very_ different options. We might add some additional flags in the future. 

Is it okay to say that we don't validate this guard if _anything_ is set?

ayende replied:
I think so.
My worry is that we may break other features because of this.
Resolving conflicts, coming from replication, etc.

The error we raise here is mostly meant to help detect race conditions, and many of those issues aren't really possible to seamlessly resolve.
I would rather have it skip in most cases, since if this happened a few seconds later, it will go through with no issue, mostly likely.

arekpalinski replied:
@karmeli87 your opinion about that?

karmeli87 replied:
Not sure to be honest... the state here might be very complex and different, so I understand the worry of breaking some features.

For fun, I would maybe try to change the default of all tx to be cluster-wide. 
Especially for tests that are testing conflicts... usually we do that with external replication, which shouldn't break anything here, I think.
I do expect some tests to fail... but most of them should pass, I think.

arekpalinski commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -54,11 +56,44 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
+                if (nonPersistentDocumentFlags != NonPersistentDocumentFlags.None) // replication or engine running an operation, we can skip checking it 
+                    return;
+
+                long indexFromOChangeVector = ChangeVectorUtils.GetEtagById(changeVector, _parent._documentDatabase.ClusterTransactionId);
```
Typo? "OChangeVector"

arekpalinski commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -54,11 +56,44 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
+                    return;
+                
+                using var _ = _serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext clusterContext);
+                clusterContext.OpenReadTransaction();
```
Missing `using`? (I know it's going to be disposed on context dispose but still prefer to be explicit)

arekpalinski commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -54,11 +56,44 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
+                using var _ = _serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext clusterContext);
+                clusterContext.OpenReadTransaction();
+                var guardId = CompareExchangeKey.GetStorageKey(_parent._documentDatabase.Name, ClusterTransactionCommand.GetAtomicGuardKey(id));
+                var (indexFromCluster, val) = _serverStore.Cluster.GetCompareExchangeValue(clusterContext, guardId);
```
Compare exchange keys are case insensitive, right?

ayende replied:
Yes, why?

arekpalinski replied:
Great. We're combining the key of a compare exchange using document IDs which are case insensitive - so just wanted to make sure that the same applies here.

arekpalinski commented on src/Raven.Server/Documents/Replication/ChangeVectorParser.cs:
```diff
@@ -7,7 +7,9 @@ namespace Raven.Server.Documents.Replication
     public static class ChangeVectorParser
     {
         public const string RaftTag = "RAFT";
+        public const string TrxnTag = "TRXN";
```
What is actually `Trxn`? Transaction compare exchange?

ayende replied:
Just an way to indicate that, yes. We can call it something else, couldn't think of anything more meaningful in 4 chars.

arekpalinski replied:
Then maybe `TXCX`? TX - from "transaction", CX - from compare exchange

ayende replied:
I don't like that, I chose `TRXN` to look like transaction.
I would rather have something meaningful, rather than abbervation

arekpalinski replied:
ok

arekpalinski commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -190,15 +217,66 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
+                    Id = atomicGuardKey,
+                    Index = changeVectorIndex,
+                    Document = context.ReadObject(new DynamicJsonValue { ["Id"] = docId }, "cmp-xchg-content"),
+                    Error = $"Guard compare exchange value '{atomicGuardKey}' index does not match the transaction index's {changeVectorIndex} change vector on {docId}"
```
Do we have to send this error already with the command data? Cannot we just recognize that somehow after the validation fails and then add / expose this error when needed?

ayende replied:
The problem is that we don't have a good place to give the _context_ for the error.
The actual validation happens later, and we may also have other compare exchange operations explicitly modified.

arekpalinski replied:
My concern was that we will be _always_ sending this error string while we already have everything - `atomicGuardKey`, `changeVectorIndex` and `docId` - inside that command to build that error message.

arekpalinski commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -190,15 +217,66 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
             var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
             var commandsCountPerDatabase = context.Transaction.InnerTransaction.ReadTree(ClusterStateMachine.TransactionCommandsCountPerDatabase);
-            var commands = SerializedDatabaseCommands.Clone(context);
+            var commands = context.ReadObject(SerializedDatabaseCommands, "serialized-tx-commands");
```
Isn't effectively the same thing here? I mean Clone() vs ReadObject()

ayende replied:
No, `Clone` is just that, doesn't promise to do more than just copy the value.
`ReadObject` will regenerate the document, which is needed since we modify the commands.

arekpalinski replied:
`Clone` will also call `ReadObject` if there are any modifications:

https://github.com/ravendb/ravendb/blob/400ef1b594b85909edfb532ef1c9c5a89e81aed5/src/Sparrow/Json/BlittableJsonReaderObject.cs#L1054-L1069

ayende replied:
Only on the first level, but we aren't modifying the first level.

arekpalinski commented on src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs:
```diff
@@ -146,10 +146,14 @@ public unsafe bool Validate(ClusterOperationContext context, Table items, long i
                 {
                     currentIndex = *(long*)reader.Read((int)ClusterStateMachine.CompareExchangeTable.Index, out var _);
+
+                    if (index == -1)
```
Can we put `-1` as a const and use it in all relevant places?

ppekrol commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -43,7 +49,7 @@ public static ClusterTransactionDataCommand FromCommandData(BatchRequestParser.C
                 {
                     Id = command.Id,
-                    ChangeVector = command.ChangeVector,
+                    ChangeVector = command.OldChangeVector ?? command.ChangeVector,
```
this is for backward compatibility? Maybe add a comment?

ayende replied:
This is actually never used anywhere else and wasn't used at all previously.

ppekrol commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -94,18 +103,22 @@ public DynamicJsonValue ToJson()
         [JsonDeserializationIgnore]
         public ClusterTransactionOptions Options;
+        
+        public bool DisableAtomicDocumentWrites;
```
On deserialization of old commands this will be 'false'. Is this OK? Shouldn't this be nullable?

ayende replied:
No, at this point, we have to make a decision. The value is already setup properly when reading the command over the network.

ppekrol replied:
No if you have commands from e.g. 4.2 in the log and you are applying them?

arekpalinski commented on src/Raven.Server/Utils/ChangeVectorUtils.cs:
```diff
@@ -216,10 +216,9 @@ public static string MergeVectors(string vectorAstring, string vectorBstring)
                 return vectorAstring;
 
-            if (_mergeVectorBuffer == null)
-                _mergeVectorBuffer = new EquatableList<ChangeVectorEntry>();
```
Was there any reason why it was `EquatableList`?

ayende replied:
No, that is why I made the change.

ppekrol commented on src/Raven.Client/Documents/Commands/Batches/DeleteCommandData.cs:
```diff
@@ -9,24 +9,38 @@ namespace Raven.Client.Documents.Commands.Batches
         {
             Id = id ?? throw new ArgumentNullException(nameof(id));
             ChangeVector = changeVector;
+            OldChangeVector = oldChangeVector;
```
I do not like the name, does not describe the purpose very well.

ayende replied:
I'm open for a new one.
Basically, this is the change vector that was read by the client, we need to send it to the server, but we can't use `ChangeVector`, since that isn't optimistic concurrency check.

ppekrol replied:
OriginalChangeVector? ClusterChangeVector? Not sure...

arekpalinski commented on test/SlowTests/Issues/RavenDB-16614.cs:
```diff
@@ -0,0 +1,329 @@
+                await Assert.ThrowsAsync<ConcurrencyException>(() => session.SaveChangesAsync());
+            }
+        }
+    }
```
Can we have some tests for import / export, backup / restore, replication?

ayende replied:
We have `WillDeleteOrphanedAtomicGuards_AfterRestoreFromBackup` for backups / restore.
Replication is tested: `CanHandleConflictsWithClusterTransactionIndex`

ppekrol commented on src/Raven.Client/Json/Serialization/NewtonsoftJson/Internal/BlittableJsonWriter.cs:
```diff
@@ -277,10 +277,13 @@ private void WritePropertyValue(string propName, object value)
                     WriteDictionary(iDictionary);
                     break;
                 case DynamicJsonValue val:
+                    _manualBlittableJsonDocumentBuilder.StartWriteObject();
```
That's surprising

ayende replied:
Yes, we never used that, it seems.

ppekrol commented on src/Raven.Server/Config/RavenConfiguration.cs:
```diff
@@ -373,7 +374,7 @@ public static RavenConfiguration CreateForDatabase(RavenConfiguration parent, st
         internal static RavenConfiguration CreateForTesting(string name, ResourceType resourceType, string customConfigPath = null)
         {
-            return new RavenConfiguration(name, resourceType, customConfigPath);
+            return new RavenConfiguration(name, resourceType, customConfigPath, skipEnvironmentVariables: true);
```
Is this necessary? I think it can cause potential issues in the future.

ayende replied:
It fails the test if you are using 

![image](https://user-images.githubusercontent.com/116915/119640159-ff83f100-be20-11eb-803a-79f2b29cae4e.png)

ppekrol replied:
FYI, a lot of tests will fail if you will set RAVEN.Indexing.Disabled to true. IMO we should revert this. This is local problem (of your machine) and now we made it a potential global problem.

ayende replied:
The problem is that this is something that we _need_ if we want to debug secured tests.
Otherwise, this creates a lot of hassle.
This function is used in 3 locations, and this is modified only for this test.

ppekrol commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -73,6 +108,12 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
             var newEtag = _documentsStorage.GenerateNextEtag();
             var modifiedTicks = _documentsStorage.GetOrCreateLastModifiedTicks(lastModifiedTicks);
+            
+            var compareClusterTransaction = new CompareClusterTransactionId(this);
```
Maybe we should avoid creating this every time? Maybe move it inside?

ayende replied:
Note that this is a struct, there isn't actually cost to create it. I separate it for code hygiene issues

ppekrol reviewed (commented):
Q:
- How this affects import/export?
- How this affects backup/restore?

karmeli87 commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -988,14 +1011,21 @@ private void PrepareForEntitiesDeletion(SaveChangesData result, IDictionary<stri
+                        if (TransactionMode == TransactionMode.ClusterWide)
+                        {
+                            // we need this to send the cluster transaction index to the cluster state machine
+                            deleteCommandData.Document = documentInfo.Metadata;
```
Since you moved this to the change vector, do we need that?

ayende replied:
No, removed.

karmeli87 commented on src/Raven.Client/Documents/Commands/Batches/DeleteCommandData.cs:
```diff
@@ -9,24 +9,38 @@ namespace Raven.Client.Documents.Commands.Batches
+            OldChangeVector = oldChangeVector;
         }
 
+        public string OldChangeVector { get; }
```
If I understand correctly, now we will always send the change vector for concurrency check against the atomic guard. 
Maybe instead, we should add `ConcurrencyCheckMode` enum which is more explicit and clear

ayende replied:
That would break older clients, however, we need to send a different value, which won't cause issues.

karmeli87 replied:
It might break when we use new client but old server. 
But the client already knows the server version `LastServerVersion`, so we can decided whether to send the change vector or not.

ayende replied:
Pretty sure that we should ignore a property that we don't know, no?
That said, a 5.2 client isn't supported for 5.0, but we'll need to check that.

ppekrol replied:
Karmel is talking about:
a) adding new property called ConcurrencyCheckMode and using that in 5.2 server along with ChangeVector instead of having OldChangeVector
b) when communicating with 5.1 or earlier server detect that and use the ChangeVector property appropriately

karmeli87 replied:
I think so as well.. 
So 5.2 with 5.2 is easy, since we will always send the `ConcurrencyCheckMode` and know what to do with that..
For 5.2 server talking with 5.0 client not sending an the `ConcurrencyCheckMode` will mean Optimistic check by default.
For 5.0 and 5.2 client, the client can recognize the server version and not send the change vector if not needed.

If feasible, I think it is a much cleaner solution

ayende replied:
Okay, sound reasonable.

karmeli87 commented on src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:
```diff
@@ -583,17 +585,96 @@ public async ValueTask DisposeAsync()
+                        await _database.TxMerger.Enqueue(updateClusterTransactionIndex);
+                        foreach (var docId in updateClusterTransactionIndex.CompareExchangeToRemove)
+                        {
+                            _compareExchangeRemoveCommands.Add(new RemoveCompareExchangeCommand(_database.Name,
```
let's use `AddOrUpdateCompareExchangeBatchCommand` it can take multiple `RemoveCommands` ( I know the command name isn't great)

ayende replied:
That is what is actually happening, we add all those commands, and then send them in a single batch on 

https://github.com/ravendb/ravendb/blob/fe20d9087cfcb71fb824014661cd2c5196b160d5/src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs#L680-L684

ayende committed:
RavenDB-16614 - Fixing PR comments

haludi commented:
What about:
Client 1 loads a document from node `A`.
Client 2 delete it in node `B`.
Client 1 modifiy the docuemtn and `SaveChanges`.

Should it fail?

```
[Fact]
public async Task ClusterWideTransaction_WhenLoadAndUpdateDeleteInParallel_ShouldSucceedOnlyInTheFirst()
{
    var (nodes, leader) = await CreateRaftCluster(3, shouldRunInMemory: false);

    using var documentStore = GetDocumentStore(new Options {Server = leader, ReplicationFactor = nodes.Count});

    using var disposable = DocumentStores(nodes, documentStore.Database, out var stores);

    var entity = new TestObj();
    using (var session = documentStore.OpenAsyncSession())
    {
        await session.StoreAsync(entity);
        await session.SaveChangesAsync();
    }

    var amre = new AsyncManualResetEvent();
    var amre2 = new AsyncManualResetEvent();
    var task = Task.Run(async () =>
    {
        using var session = stores[0].OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide });
        var loaded = await session.LoadAsync<TestObj>(entity.Id);
        amre.Set();

        loaded.Prop = "Changed";

        await amre2.WaitAsync();
        await Assert.ThrowsAnyAsync<ConcurrencyException>(() => session.SaveChangesAsync());
    });
    await amre.WaitAsync();
    using (var session = stores[1].OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide }))
    {
        var loaded = await session.LoadAsync<TestObj>(entity.Id);
        session.Delete(loaded);
        await session.SaveChangesAsync();

        amre2.Set();
    }

    await task;
}
```

ayende commented:
I don't think so.
What we are doing is asserting that there is no guard for this, and there
isn't.
The 2nd transaction is a cluster wide, but the document wasn't part of a
cluster wide transaction previously.
In other words, there wasn't a guard before the operation.

If you make the 1st transaction cluster wide, it should fail.

On Thu, May 27, 2021 at 10:20 AM Igal Merhavia ***@***.***>
wrote:

> What about:
> Client 1 loads a document from node A.
> Client 2 delete it in node B.
> Client 1 modifiy the docuemtn and SaveChanges.
>
> Should it fail?
>
> [Fact]
> public async Task ClusterWideTransaction_WhenLoadAndUpdateDeleteInParallel_ShouldSucceedOnlyInTheFirst()
> {
>     var (nodes, leader) = await CreateRaftCluster(3, shouldRunInMemory: false);
>
>     using var documentStore = GetDocumentStore(new Options {Server = leader, ReplicationFactor = nodes.Count});
>
>     using var disposable = DocumentStores(nodes, documentStore.Database, out var stores);
>
>     var entity = new TestObj();
>     using (var session = documentStore.OpenAsyncSession())
>     {
>         await session.StoreAsync(entity);
>         await session.SaveChangesAsync();
>     }
>
>     var amre = new AsyncManualResetEvent();
>     var amre2 = new AsyncManualResetEvent();
>     var task = Task.Run(async () =>
>     {
>         using var session = stores[0].OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide });
>         var loaded = await session.LoadAsync<TestObj>(entity.Id);
>         amre.Set();
>
>         loaded.Prop = "Changed";
>
>         await amre2.WaitAsync();
>         await Assert.ThrowsAnyAsync<ConcurrencyException>(() => session.SaveChangesAsync());
>     });
>     await amre.WaitAsync();
>     using (var session = stores[1].OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide }))
>     {
>         var loaded = await session.LoadAsync<TestObj>(entity.Id);
>         session.Delete(loaded);
>         await session.SaveChangesAsync();
>
>         amre2.Set();
>     }
>
>     await task;
> }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ravendb/ravendb/pull/12243#issuecomment-849398932>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4RM2XZCALTQALZPVAJMLTPXXE7ANCNFSM45M374LA>
> .
>


-- 
<https://ravendb.net/> *Oren Eini*
*CEO   /   Hibernating Rhinos LTD <https://hibernatingrhinos.com/>*
Mobile:  972-52-548-6969
Sales:  ***@***.***
Skype:  ayenderahien
Support:  ***@***.***
<https://www.facebook.com/pages/RavenDB/265907650186374>
<https://twitter.com/ravendb>
<https://www.linkedin.com/company/hibernating-rhinos-ltd-/>
<https://ravendb.net/emailsignature/displayeventpage>

haludi committed:
RavenDB-16614 Test for "Atomic read write in cluster wide transaction without conflicts"

ayende committed:
RavenDB-16614
- Support older servers from 5.2 client
- Older commands running on new servers will no apply atomic guard operations

ayende committed:
RavenDB-16614 - Fixing the build

ppekrol commented on src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs:
```diff
@@ -15,37 +15,46 @@ namespace Raven.Client.Documents.Commands.Batches
+        private bool? _supportsAtomicWrites;
         private readonly List<Stream> _attachmentStreams;
         private readonly HashSet<Stream> _uniqueAttachmentStreams;
+        private readonly JsonOperationContext _context;
```
not used?

karmeli87 replied:
removed

ppekrol commented on src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs:
```diff
@@ -15,37 +15,46 @@ namespace Raven.Client.Documents.Commands.Batches
         private readonly TransactionMode _mode;
 
-        public SingleNodeBatchCommand(DocumentConventions conventions, JsonOperationContext context, IList<ICommandData> commands, BatchOptions options = null, TransactionMode mode = TransactionMode.SingleNode)
+        public SingleNodeBatchCommand(DocumentConventions conventions,  IList<ICommandData> commands, BatchOptions options = null, TransactionMode mode = TransactionMode.SingleNode)
```
breaking change - we need to revert this

karmeli87 replied:
done

ppekrol commented on src/Raven.Client/Documents/Commands/Batches/DeleteCommandData.cs:
```diff
@@ -9,24 +9,38 @@ namespace Raven.Client.Documents.Commands.Batches
+            OriginalChangeVector = originalChangeVector;
         }
 
+        public string OriginalChangeVector { get; }
```
Can we use ChangeVector property and add e.g. ConcurrencyCheckMode enum property here?

ppekrol commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -69,6 +76,7 @@ public class ClusterTransactionOptions : IDynamicJson
             public TimeSpan? WaitForIndexesTimeout;
             public bool WaitForIndexThrow;
             public List<string> SpecifiedIndexesQueryString;
+            public bool DisableAtomicDocumentWrites;
```
Double-check. For old commands (e.g. from 4.2 era) it will be set to false when Cluster Log is applied. We want that?

ayende committed:
RavenDB-16614 Fixing tests

haludi committed:
RavenDB-16614 Use `ClusterTransactionCommand` in import to database destination

haludi committed:
RavenDB-16614 Don't add compare exchage if there is no document & batch consider _clusterTransactionCommands

haludi committed:
RavenDB-16614 Import with cluster wide trx get only data

karmeli87 committed:
RavenDB-16614 mixed cluster should retain the old behaviour

karmeli87 committed:
RavenDB-16614 update correctly the database change vector

karmeli87 committed:
RavenDB-16614 address PR comments

karmeli87 committed:
RavenDB-16614 remove TRXN for new documents

karmeli87 committed:
RavenDB-16614 separate compre exchange from cluster wide tx on import

haludi committed:
RavenDB-16614 ClusterTransactionCommand - move cluster version check to command creation

karmeli87 committed:
RavenDB-16614 TRXN Tag should be only on the document's change vector

karmeli87 committed:
RavenDB-16614 fallback to true for DisableAtomicDocumentWrites for < 5.2 clients

karmeli87 committed:
RavenDB-16614 remove redundant DisableAtomicDocumentWrites

it is enough to have it once on the options.

ppekrol commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -129,6 +136,21 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
+                return;
+            }
+
+            if (clientVersion.Major <= 5 && clientVersion.Minor < 2)
```
`clientVersion.Major < 5 || (clientVersion.Major == 5 && clientVersion.Minor < 2)`

would be the most appropriate one?

karmeli87 replied:
done

karmeli87 committed:
RavenDB-16614 fix client version check

haludi committed:
RavenDB-16614 Use local doc change vector for OriginalChangeVector instead of empty while importing

karmeli87 committed:
RavenDB-16614 set ClusterTransactionId & DatabaseGroupId for restoring
